src/h5web/providers/models.ts

```diff
@@ -60,10 +61,11 @@ type PrimitiveType<T extends HDF5Type> = T extends HDF5NumericType
   ? Complex
   : unknown;
 
-export type Value<
-  S extends HDF5Shape,
-  T extends HDF5Type
-> = S extends HDF5SimpleShape ? PrimitiveType<T>[] : PrimitiveType<T>;
+export type Value<D extends Dataset> = D['shape'] extends HDF5SimpleShape
+  ? PrimitiveType<D['type']>[]
+  : D['shape'] extends HDF5ScalarShape
+  ? PrimitiveType<D['type']>
+  : never;
```

Comment:
`D['shape']` and `D['type']` are used to access the TS types of the dataset's HDF5 shape and type.

Note that I added a condition on `HDF5ScalarShape` with a fallback to deal with the `null` shape. We won't technically need it, but it feels more robust and complete.